_delphi_utils_python/delphi_utils/logger.py

```diff
@@ -84,10 +119,116 @@ def get_structured_logger(name=__name__,
     system_logger = logging.getLogger(name)
     if filename and not system_logger.handlers:
         system_logger.addHandler(logging.FileHandler(filename))
-    system_logger.setLevel(logging.INFO)
+    system_logger.setLevel(log_level)
     logger = structlog.wrap_logger(system_logger)
 
     if log_exceptions:
         handle_exceptions(logger)
 
     return logger
+
+
+
+
+class LoggerThread():
+    """
+    the bare structlog loggers are thread-safe but not multiprocessing-safe.
+    a `LoggerThread` will spawn a thread that listens to a mp.Queue
+    and logs messages from it with the provided logger,
+    so other processes can send logging messages to it
+    via the logger-like `SubLogger` interface.
+    the SubLogger even logs the pid of the caller.
+
+    this is good to use with a set of jobs that are part of a mp.Pool,
+    but isnt recommended for general use
+    because of overhead from threading and multiprocessing,
+    and because it might introduce lag to log messages.
+    """
+
+    class SubLogger():
+        """
+        Multiprocessing-safe logger-like interface
+        to convey log messages to a listening LoggerThread.
+        """
+
+        def __init__(self, queue):
+            self.queue = queue
+
+        def _log(self, level, *args, **kwargs):
+            kwargs_plus = {'sub_pid': multiprocessing.current_process().pid}
+            kwargs_plus.update(kwargs)
+            self.queue.put([level, args, kwargs_plus])
+
+        def info(self, *args, **kwargs):
+            """Log an INFO level message."""
+            self._log(logging.INFO, *args, **kwargs)
+
+        def warn(self, *args, **kwargs):
+            """Log a WARN level message."""
+            self._log(logging.WARN, *args, **kwargs)
+
+
+    def get_sublogger(self):
+        """Accessor method to retrieve a SubLogger for this LoggerThread."""
+        return self.sublogger
+
+    def __init__(self, logger, q=None):
+        self.logger = logger
+        if q:
+            self.msg_queue = q
+        else:
+            self.msg_queue = multiprocessing.Queue()
+
+        def logger_thread_worker():
+            logger.info('thread started')
+            while True:
+                msg = self.msg_queue.get()
+                if msg == 'STOP':
+                    logger.debug('received stop signal')
+                    break
+                level, args, kwargs = msg
+                if level == logging.INFO:
+                    logger.info(*args, **kwargs)
+                if level == logging.WARN:
+                    logger.warn(*args, **kwargs)
+                else:
+                    logger.error('received unknown logging level!  exiting...')
+                    break
+            logger.debug('stopping thread')
+
+        self.thread = threading.Thread(target=logger_thread_worker,
+                                       name="LoggerThread__"+logger.name)
```

Comment:
I don't quite follow the purpose of this thread... it seems like it should interact with self.sublogger in some way, no? Could you expand a bit?

Reply:
the thread doesnt really "interact" with the sublogger; the sublogger acts indirectly upon the thread.  the thread is spawned in the background with a handle to a queue and a handle to the provided Logger, and it just logs anything that shows up on the queue.  the sublogger is used by things outside the thread (or even outside the process!) to enqueue things to be logged.